### ISSUE TITLE
NativeImageLoader performance

### ISSUE THREAD

**Issue by crockpotveggies** [deeplearning4j/DataVec#232]
#### Issue Description

After wrapping NativeImageLoader.asMatrix() on a long-running training instance, image loading is taking between `.5` milliseconds to `1.7ms` *per image*, with times as high as `17ms` being seen every 100 loads or so. When running large batch sizes, this adds up.

For example, a batch size of 72 will take RecordReaderDataSetIterator an average of `` when calling `next()`.

After some investigation, I believe the best place to parallelize this operation is in RecordReaderDataSetIterator and add multiple threads for handling the internal record reader next(). Alternatively, it may be easier to put multiple threads inside ImageRecordReader itself.

I imagine an API change here may be necessary.

This Gist shows output timings of both metrics mentioned above on a warmed up JVM.

https://gist.github.com/crockpotveggies/aa81d9d37fd84b88a1de66bd38a01581

#### Contributing

I'd like to take a look at implementing this.

**Comment by crockpotveggies** [comment]
cc @raver119

**Comment by saudet** [comment]
Sounds good to me 👍

**Comment by raver119** [comment]
Waste of time then. If single batch load time is smaller then single iteration time - your efforts won't have any effect on training time. Amdahl's law.

Check iterator.next() time within MLN/CG fit() instead, as discussed before. That should be more important given your new timings.

**Comment by raver119** [comment]
P.s. however, for primitive things like LeNet - there might be positive effect, obviously. It's all about iteration time vs ETL time.

**Comment by raver119** [comment]
>After some investigation, I believe the best place to parallelize this operation is in RecordReaderDataSetIterator and add multiple threads for handling the internal record reader next(). Alternatively, it may be easier to put multiple threads inside ImageRecordReader itself.

Nope, as we've discussed before - only the first option is viable. Basically we HAVE to keep IO sequential for many RecordReaders out there. Who knows, where data comes from. But internal conversion into INDArrays - that's what we can definitely do in parallel.

**Comment by raver119** [comment]
I've added ETL timing output to PerformanceListener.

**Comment by raver119** [comment]
That's ETL timing without ADSI, LeNet example:

```
o.d.o.l.PerformanceListener - Device: [0]; ETL: 1 ms; iteration 301; iteration time: 13 ms; samples/sec: 4923.077; batches/sec: 76.923; score: 0.13040581776170682;
o.d.o.l.PerformanceListener - Device: [0]; ETL: 1 ms; iteration 351; iteration time: 13 ms; samples/sec: 4923.077; batches/sec: 76.923; score: 0.14284975756572094;
o.d.o.l.PerformanceListener - Device: [0]; ETL: 0 ms; iteration 401; iteration time: 12 ms; samples/sec: 5333.333; batches/sec: 83.333; score: 0.10325338956317653;
o.d.o.l.PerformanceListener - Device: [0]; ETL: 2 ms; iteration 451; iteration time: 14 ms; samples/sec: 4571.428; batches/sec: 71.429; score: 0.07725700314539087;
o.d.o.l.PerformanceListener - Device: [0]; ETL: 1 ms; iteration 501; iteration time: 12 ms; samples/sec: 5333.333; batches/sec: 83.333; score: 0.10587001916430576;
```

That's ETL timing with ADSI, same LeNet example:
```
o.d.o.l.PerformanceListener - Device: [0]; ETL: 0 ms; iteration 151; iteration time: 12 ms; samples/sec: 5333.333; batches/sec: 83.333; score: 0.23935580142267293;
o.d.o.l.PerformanceListener - Device: [0]; ETL: 0 ms; iteration 201; iteration time: 12 ms; samples/sec: 5333.333; batches/sec: 83.333; score: 0.26675685531985716;
o.d.o.l.PerformanceListener - Device: [0]; ETL: 0 ms; iteration 251; iteration time: 12 ms; samples/sec: 5333.333; batches/sec: 83.333; score: 0.1693427142217654;
o.d.o.l.PerformanceListener - Device: [0]; ETL: 0 ms; iteration 301; iteration time: 11 ms; samples/sec: 5818.182; batches/sec: 90.909; score: 0.1761187335082122;
o.d.o.l.PerformanceListener - Device: [0]; ETL: 0 ms; iteration 351; iteration time: 12 ms; samples/sec: 5333.333; batches/sec: 83.333; score: 0.1705211329791694;
o.d.o.l.PerformanceListener - Device: [0]; ETL: 0 ms; iteration 401; iteration time: 12 ms; samples/sec: 5333.333; batches/sec: 83.333; score: 0.2994427617451107;
```

**Comment by raver119** [comment]
Yesterday i've spent some time looking into parallelism at IRR & RRDSI, and have to admit now: we might need some parallelism at ETL level.

We might need it for efficient data parallelism. As number of devices grows, single-threded prefetcher at ADSI might become inefficient, due to current ParallelWrapper algorithm:

- ADSI prefetches X datasets.
- ParallelWrapper fetches one DataSet, and passes it into trainer thread.
- Without any block, PW main thread cycles back to prefetcher, and tries to get one more DataSet.
- This cycle continues up to number of trainer threads active, and blocks until all threads had finished their job. But:
- While main thread was cycling through feed loop, trainer threads were actually training. So there are chances that early trainer threads already finished training at this moment of time.
- And here starts IO starvation:
- If ADSI wasn't 100% recovered and isn't filled with DataSets, each iterator.next() will be taking more time then expecting, introducing more and more desync between trainer threads.

That's theory atm, but today i'll write simulation for it.

**Comment by raver119** [comment]
Also, as discussed with @crockpotveggies we introduce issues with Workspaces here when we add parallelism: workspaces are discoverable within same thread, so once we add parallelism to ETL pipeline, we won't be able to benefit from automatic workspace propagation. And we'll need to propagate them ourself.

**Comment by raver119** [comment]
Damn, everything is just AWFUL. Details coming soon, sure. But now i think we have no choice here... ETL overhaul is imminent.

**Comment by raver119** [comment]
```
o.d.e.LinearSimulator - Linear scaling simulation: 25 ms per dataset, 100 ms training time, 1000 total examples: -----------------------------
o.d.e.Simulator - Throughput for 1 workers: 10 datasets/second

o.d.p.ParallelWrapper - p85 ETL time: 0 ms; p50 ETL time: 0 ms
o.d.e.Simulator - Throughput for 2 workers: 19 datasets/second
o.d.p.ParallelWrapper - p85 ETL time: 0 ms; p50 ETL time: 0 ms
o.d.e.Simulator - Throughput for 4 workers: 40 datasets/second
o.d.p.ParallelWrapper - p85 ETL time: 25 ms; p50 ETL time: 0 ms
o.d.e.Simulator - Throughput for 6 workers: 40 datasets/second
o.d.p.ParallelWrapper - p85 ETL time: 25 ms; p50 ETL time: 23 ms
o.d.e.Simulator - Throughput for 8 workers: 40 datasets/second
```

iteration time for 1 device / time per ds = number of optimal workers. for higher number of threads, single-threaded prefetcher won't be able to fulfill all workers without delays.

**Comment by raver119** [comment]
Simulator can be found here: https://github.com/deeplearning4j/ETL-benchmark/blob/801864eaadbff393751ca793c787ea32e5d23fdd/src/main/java/org/deeplearning4j/etl/LinearSimulator.java

**Comment by huitseeker** [comment]
@raver119 any news summarizing how awful things are here ?
